README.md
# Refraction2

`refraction2` is a Rack 2.2+ compatible re-release and drop-in replacement for
[joshsusser/refraction](https://github.com/joshsusser/refraction).

For differences from the original, see [CHANGELOG.md](./CHANGELOG.md).

## Original README

Refraction is a Rack middleware replacement for `mod_rewrite`. It can rewrite URLs before they are
processed by your web application, and can redirect using 301 and 302 status codes. Refraction is
thread-safe, so it doesn't need to be guarded by Rack::Lock.

The best thing about Refraction is that rewrite rules are written in plain old Ruby code, not some
funky web server config syntax. That means you can use Ruby regular expressions, case statements,
conditionals, and whatever else you feel like.

For example:

    Refraction.configure do |req|
      feedburner  = "http://feeds.pivotallabs.com/pivotallabs"

      if req.env['HTTP_USER_AGENT'] !~ /FeedBurner|FeedValidator/ && req.host =~ /pivotallabs\.com/
        case req.path
        when %r{^/(talks|blabs|blog)\.(atom|rss)$}        ; req.found! "#{feedburner}/#{$1}.#{$2}"
        when %r{^/users/(chris|edward)/blog\.(atom|rss)$} ; req.found! "#{feedburner}/#{$1}.#{$2}"
        end
      else
        case req.host
        when 'tweed.pivotallabs.com'
          req.rewrite! "http://pivotallabs.com/tweed#{req.path}"
        when /([-\w]+\.)?pivotallabs\.com/
          # passthrough with no change
        else  # wildcard domains (e.g. pivotalabs.com)
          req.permanent! :host => "pivotallabs.com"
        end
      end
    end

Notice the use of regular expressions, the $1, $2, etc pseudo-variables, and string interpolation.
This is an easy way to match URL patterns and assemble the new URL based on what was matched.

## Installation (Rails)

Refraction can be installed in a Rails application as a plugin.

    $ script/plugin install git://github.com/pivotal/refraction.git

It can also be used as a gem:

    $ gem install refraction

In `environments/production.rb`, add Refraction at or near the top of your middleware stack.

    config.middleware.insert_before(::Rack::Lock, ::Refraction, {})

You may want to occasionally turn on Refraction in the development environment for testing
purposes, but if your rules redirect to other servers that can be a problem.

Put your rules in `config/initializers/refraction_rules.rb` (see example above). The file name
doesn't actually matter, but convention is useful.

## Server Configuration

If your application is serving multiple virtual hosts, it's probably easiest to configure your web
server to handle a wildcard server name and let Refraction handle managing the virtual hosts. For
example, in nginx, that is done with a `server_name _;` directive.

## Writing Rules

Set up your rewrite/redirection rules during your app initialization using `Refraction.configure`.
The `configure` method takes a block which is run for every request to process the rules. The block
is passed a RequestContext object that contains information about the request URL and environment.
The request object also has a small API for effecting rewrites and redirects.

> Important note: don't do a `return` from within the configuration
> block.  That would be bad (meaning your entire application would
> break).  That's just how blocks work in Ruby.

### `RequestContext#set(options)`

The `set` method takes an options hash that sets pieces of the rewritten URL or redirect location
header.

  * :scheme - Usually `http` or `https`.
  * :host - The server name.
  * :port - The server port.  Usually not needed, as the scheme implies a default value.
  * :path - The path of the URL.
  * :query - Added at the end of the URL after a question mark (?)

Any URL components not explicitly set remain unchanged from the original request URL. You can use
`set` before calls to `rewrite!`, `permanent!`, or `found!` to set common values. Subsequent
methods will merge their component values into values from `set`.

### `RequestContext#rewrite!(options)`

The `rewrite!` method modifies the request URL and relevant pieces of the environment. When
Refraction rule processing results in a `rewrite!`, the request is passed on down the Rack stack
to the app or the next middleware component. `rewrite!` can take a single argument, either an
options hash that uses the same options as the `set` method, or a string that sets all components
of the URL.

### `RequestContext#permanent!(options)`

The `permanent!` method tells Refraction to return a response with a `301 Moved Permanently`
status, and sets the URL for the Location header. Like `rewrite!` it can take either a string or
hash argument to set the URL or some of its components.

### `RequestContext#found!(options)`

The `found!` method tells Refraction to return a response with a `302 Found` status, and sets the
URL for the Location header. Like `#rewrite!` it can take either a string or hash argument to set
the URL or some of its components.

### `RequestContext#respond!(status, headers, content)`

Use `respond!` to return an arbitrary response to the request. This is useful for responding with
the contents of a static file. For example:

    req.respond!(503, {'Content-Type' => 'text/html'}, File.read(MAINTENANCE_PATH))

The args are largely the same as the contents of a standard Rack response array with the exceptions
that you don't need to wrap the content in an array, and the Content-Length header is generated so it
should not be supplied.

### URL components

The request object provides the following components of the URL for matching requests: `scheme`,
`host`, `port`, `path`, and `query`. It also provides a full environment hash as the `env`
attribute. For example, `req.env['HTTP_USER_AGENT']` can be used to access the request's user
agent property.

## Contributors

  * Josh Susser (maintainer)
  * Sam Pierson
  * Wai Lun Mang
  * Brian Morearty

Copyright (c) 2009-2010 Pivotal Labs. This software is licensed under the MIT License.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.1] - 2022-05-30 (first release as `refraction2`)

### Fixed

- fixed host and port overrides for Rack 2.2+

## [0.1.3] - 2010-02-01 (last release as `refraction` by Josh Susser)


spec/spec_helper.rb
require 'refraction'
require 'rack/test'


spec/refraction_spec.rb
describe Refraction do
  let(:app) { double('app') }

  describe "if no rules have been configured" do
    before do
      Refraction.configure
    end

    it "does nothing" do
      env = Rack::MockRequest.env_for('http://bar.com/about', :method => 'get')
      expect(app).to receive(:call) do |resp|
        expect(resp['rack.url_scheme']).to eq 'http'
        expect(resp['SERVER_NAME']).to eq 'bar.com'
        expect(resp['PATH_INFO']).to eq '/about'
        [200, {}, ["body"]]
      end
      response = Refraction.new(app).call(env)
    end
  end

  describe "path" do
    before do
      Refraction.configure do |req|
        if req.path == '/'
          req.permanent! 'http://yes.com/'
        elsif req.path == ''
          req.permanent! 'http://no.com/'
        end
      end
    end

    it "should be set to / if empty" do
      env = Rack::MockRequest.env_for('http://bar.com', :method => 'get')
      env['PATH_INFO'] = '/'
      response = Refraction.new(app).call(env)
      expect(response[0]).to eq 301
      expect(response[1]['Location']).to eq "http://yes.com/"
    end
  end

  describe "permanent redirection" do
    describe "using string arguments" do
      before do
        Refraction.configure do |req|
          req.permanent! "http://foo.com/bar?baz"
        end
      end

      it "should redirect everything to foo.com" do
        env = Rack::MockRequest.env_for('http://bar.com', :method => 'get')
        response = Refraction.new(app).call(env)
        expect(response[0]).to eq 301
        expect(response[1]['Location']).to eq "http://foo.com/bar?baz"
      end
    end

    describe "using hash arguments" do
      before do
        Refraction.configure do |req|
          req.permanent! :host => "foo.com", :path => "/bar", :query => "baz"
        end
      end

      it "should redirect http://bar.com to http://foo.com" do
        env = Rack::MockRequest.env_for('http://bar.com', :method => 'get')
        response = Refraction.new(app).call(env)
        expect(response[0]).to eq 301
        expect(response[1]['Location']).to eq "http://foo.com/bar?baz"
      end

      it "should redirect https://bar.com to https://foo.com" do
        env = Rack::MockRequest.env_for('https://bar.com', :method => 'get')
        response = Refraction.new(app).call(env)
        expect(response[0]).to eq 301
        expect(response[1]['Location']).to eq "https://foo.com/bar?baz"
      end

      it "should clear the port unless set explicitly" do
        env = Rack::MockRequest.env_for('http://bar.com:3000/', :method => 'get')
        response = Refraction.new(app).call(env)
        expect(response[0]).to eq 301
        expect(response[1]['Location']).to eq "http://foo.com/bar?baz"
      end
    end

    describe "using hash arguments but not changing scheme, host, or port" do
      before do
        Refraction.configure do |req|
          req.permanent! :path => "/bar", :query => "baz"
        end
      end

      it "should not clear the port" do
        env = Rack::MockRequest.env_for('http://bar.com:3000/', :method => 'get')
        response = Refraction.new(app).call(env)
        expect(response[0]).to eq 301
        expect(response[1]['Location']).to eq "http://bar.com:3000/bar?baz"
      end
    end

    describe "with or without port number" do
      before(:each) do
        Refraction.configure do |req|
          case req.host
          when "asterix.example.com"
            req.permanent! :path => "/potion#{req.path}"
          when "obelix.example.com"
            req.permanent! :host => "menhir.example.com"
          when "getafix.example.com"
            req.permanent! :scheme => "https"
          when "dogmatix.example.com"
            req.permanent! :port => 3001
          end
        end
      end

      it "should include port in Location if request had a port and didn't change scheme, host, or port" do
        env = Rack::MockRequest.env_for('http://asterix.example.com:3000/1', :method => 'get')
        response = Refraction.new(app).call(env)
        expect(response[0]).to eq 301
        expect(response[1]['Location']).to include("asterix.example.com:3000")
      end

      it "should not include port in Location if request didn't specify a port" do
        env = Rack::MockRequest.env_for('http://asterix.example.com/1', :method => 'get')
        response = Refraction.new(app).call(env)
        expect(response[0]).to eq 301
        expect(response[1]['Location']).to include("asterix.example.com")
        expect(response[1]['Location']).not_to include(":3000")
      end

      it "should remove port from Location if host was changed" do
        env = Rack::MockRequest.env_for('http://obelix.example.com:3000/1', :method => 'get')
        response = Refraction.new(app).call(env)
        expect(response[0]).to eq 301
        expect(response[1]['Location']).to include("menhir.example.com")
        expect(response[1]['Location']).not_to include(":3000")
      end

      it "should remove port from Location if scheme was changed" do
        env = Rack::MockRequest.env_for('http://getafix.example.com:3000/1', :method => 'get')
        response = Refraction.new(app).call(env)
        expect(response[0]).to eq 301
        expect(response[1]['Location']).to include("getafix.example.com")
        expect(response[1]['Location']).not_to include(":3000")
      end

      it "should change port in Location if port was changed" do
        env = Rack::MockRequest.env_for('http://dogmatix.example.com:3000/1', :method => 'get')
        response = Refraction.new(app).call(env)
        expect(response[0]).to eq 301
        expect(response[1]['Location']).to include("dogmatix.example.com:3001")
      end
    end
  end

  describe "temporary redirect for found" do
    before(:each) do
      Refraction.configure do |req|
        if req.path =~ %r{^/users/(josh|edward)/blog\.(atom|rss)$}
          req.found! "http://feeds.pivotallabs.com/pivotallabs/#{$1}.#{$2}"
        end
      end
    end

    it "should temporarily redirect to feedburner.com" do
      env = Rack::MockRequest.env_for('http://bar.com/users/josh/blog.atom', :method => 'get')
      response = Refraction.new(app).call(env)
      expect(response[0]).to eq 302
      expect(response[1]['Location']).to eq "http://feeds.pivotallabs.com/pivotallabs/josh.atom"
    end

    it "should not redirect when no match" do
      env = Rack::MockRequest.env_for('http://bar.com/users/sam/blog.rss', :method => 'get')
      expect(app).to receive(:call) do |resp|
        expect(resp['rack.url_scheme']).to eq 'http'
        expect(resp['SERVER_NAME']).to eq 'bar.com'
        expect(resp['PATH_INFO']).to eq '/users/sam/blog.rss'
        [200, {}, ["body"]]
      end
      response = Refraction.new(app).call(env)
    end
  end

  describe "rewrite url" do
    before(:each) do
      Refraction.configure do |req|
        if req.host =~ /(tweed|pockets)\.example\.com/
          req.rewrite! :host => 'example.com', :path => "/#{$1}#{req.path == '/' ? '' : req.path}"
        end
      end
    end

    it "should rewrite subdomain to scope the path for matching subdomains" do
      env = Rack::MockRequest.env_for('http://tweed.example.com', :method => 'get')
      expect(app).to receive(:call) do |resp|
        expect(resp['rack.url_scheme']).to eq 'http'
        expect(resp['SERVER_NAME']).to eq 'example.com'
        expect(resp['PATH_INFO']).to eq '/tweed'
        [200, {}, ["body"]]
      end
      Refraction.new(app).call(env)
    end

    it "should not rewrite if the subdomain does not match" do
      env = Rack::MockRequest.env_for('http://foo.example.com', :method => 'get')
      expect(app).to receive(:call) do |resp|
        expect(resp['rack.url_scheme']).to eq 'http'
        expect(resp['SERVER_NAME']).to eq 'foo.example.com'
        expect(resp['PATH_INFO']).to eq '/'
        [200, {}, ["body"]]
      end
      Refraction.new(app).call(env)
    end
  end

  describe "generate arbitrary response" do
    before(:each) do
      Refraction.configure do |req|
        req.respond!(503, {'Content-Type' => 'text/plain'}, "Site down for maintenance.")
      end
    end

    it "should respond with status, headers and content" do
      env = Rack::MockRequest.env_for('http://example.com', :method => 'get')
      response = Refraction.new(app).call(env)
      expect(response[0]).to eq 503
      expect(response[1]['Content-Length'].to_i).to eq "Site down for maintenance.".length
      expect(response[1]['Content-Type']).to eq "text/plain"
      expect(response[2]).to eq ["Site down for maintenance."]
    end
  end

  describe "environment" do
    before(:each) do
      Refraction.configure do |req|
        if req.env['HTTP_USER_AGENT'] =~ /FeedBurner/
          req.permanent! "http://yes.com/"
        else
          req.permanent! "http://no.com/"
        end
      end
    end

    it "should expose environment settings" do
      env = Rack::MockRequest.env_for('http://foo.com/', :method => 'get')
      env['HTTP_USER_AGENT'] = 'FeedBurner'
      response = Refraction.new(app).call(env)
      expect(response[0]).to eq 301
      expect(response[1]['Location']).to eq "http://yes.com/"
    end
  end
end


lib/refraction.rb
require 'rack'
require_relative 'refraction/version'

class Refraction
  class Request < Rack::Request
    attr_reader :action, :status, :message

    # backward compatibility: support URI::HTTP component names
    def method; request_method; end
    def query;  query_string;   end

    ### actions

    def set(options)
      if options.is_a?(String)
        @re_location = options
      else
        @re_scheme = options[:protocol] if options[:protocol] # :protocol is alias for :scheme
        @re_scheme = options[:scheme]   if options[:scheme]
        @re_host   = options[:host]     if options[:host]
        @re_port   = options[:port]     if options[:port]
        @re_path   = options[:path]     if options[:path]
        @re_query  = options[:query]    if options[:query]
      end
    end

    def rewrite!(options)
      @action = :rewrite
      set(options)
    end

    def permanent!(options)
      @action = :permanent
      @status = 301
      set(options)
      @message = "moved to #{@uri}"
    end

    def found!(options)
      @action = :found
      @status = 302
      set(options)
      @message = "moved to #{@uri}"
    end

    def respond!(status, headers, content)
      @action = :respond
      @status = status
      @headers = headers
      @message = content
    end

    ### response

    def response
      headers = @headers || { 'Location' => location, 'Content-Type' => 'text/plain' }
      headers['Content-Length'] = message.length.to_s
      [status, headers, [message]]
    end

    def location
      @re_location || url
    end

    # use original request's values when not set explicitly
    def scheme;       @re_scheme || super; end
    def host;         @re_host   || super; end
    def path;         @re_path   || super; end
    def query_string; @re_query  || super; end

    # changing the scheme or host means use default port instead of port in original request
    def port
      @re_port || ((@re_scheme || @re_host) && DEFAULT_PORTS[scheme]) || super
    end

    def host_with_port
      port.nil? || port == DEFAULT_PORTS[scheme] ? host : "#{host}:#{port}"
    end

    def http_host
      port ? "#{host}:#{port}" : host
    end
  end ### class Request

  def self.configure(&block)
    @rules = block
  end

  def self.rules
    @rules
  end

  def initialize(app)
    @app = app
  end

  def rules
    self.class.rules
  end

  def call(env)
    return @app.call(env) unless rules

    request = Request.new(env)
    rules.call(request)

    case request.action
    when :permanent, :found, :respond
      request.response
    when :rewrite
      env["rack.url_scheme"]  = request.scheme
      env["HTTP_HOST"]        = request.http_host
      env["SERVER_NAME"]      = request.host
      env["HTTP_PORT"]        = request.port if request.port
      env["PATH_INFO"]        = request.path
      env["QUERY_STRING"]     = request.query
      env["REQUEST_URI"]      = request.fullpath
      @app.call(env)
    else
      @app.call(env)
    end
  end
end
